BootloaderCorePkg/Include/Library/SocInitLib.h

```diff
@@ -7,6 +7,110 @@
 
 #ifndef _SOC_INIT_LIB_H_
 
+#define S3_SAVE_REG_SIG     SIGNATURE_32('S', '3', 'S', 'R')
+
+//
+// Format to share info between bootloader and payload.
+// This includes SMM Base info to relocate bases
+// on S3 resume, and RegInfo to restore certain
+// register states on S3 resume path.
+// 'RegInfo' (except value of reg) part is populated by the bootloader,
+// 'SmmBase' part is populated by the payload
+//
+
+      //  -------------------------- <----TSEG_BASE------
+      //  |     Signature (SCOM)   |                    ^
+      //  --------------------------                    |
+      //  |         CpuCnt         |                    |
+      //  --------------------------                    |
+      //  |         Unused         |                    |
+      //  --------------------------                    |
+      //  |          Rsvd          |                    |
+      //  --------------------------               SMMBASE_INFO
+      //  |       SmmBase[0]       |                    |
+      //  --------------------------                    |
+      //  |       SmmBase[1]       |                    |
+      //  --------------------------                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  --------------------------                    |
+      //  |       SmmBase[n]       |                    V
+      //  -----------------------------------------------
+      //  |     Signature (S3SR)   |                    ^
+      //  --------------------------                    |
+      //  |         RegCnt         |                    |
+      //  --------------------------                    |
+      //  |       RegInfo[0]       |                    |
+      //  --------------------------                    |
+      //  |       RegInfo[1]       |            S3_SAVE_REG_INFO
+      //  --------------------------                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  |         .....          |                    |
+      //  --------------------------                    |
+      //  |       RegInfo[n]       |                    V
+      //  -----------------------------------------------
+
+
+#pragma pack(1)
+
+typedef struct {
+  UINT8         Type;
+  UINT8         Width;
+  UINT8         Rsvd[2];
+  UINT32        Addr;
+  UINT32        Val;
+} REG_INFO;
+
+typedef struct {
+  UINT32        Signature;
+  UINT32        RegCnt;
+  REG_INFO      RegInfo[];
+} S3_SAVE_REG;

```

Comment:
It is not consistent by defining RegCnt UINT32 and CpuCnt UINT8.
It would be easier to parse a structure using a common header for all the structures in S3 SAVE area.
e.g.:
typedef struct {
  UINT32        Signature;
  UINT16        Count;
  UINT16        TotalSize;
} S3_SAVE_HEADER;

Reply:
Makes sense, will make a common header for S3_SAVE_REG and SMMBASE_INFO strcutres.